crates/matrix-sdk-crypto/src/olm/utility.rs

```diff
@@ -50,56 +53,148 @@ impl SignJson for Ed25519SecretKey {
     }
 }
 
+/// A trait for objects that are able to verify signatures.
 pub trait VerifyJson {
-    /// Verify a signed JSON object.
+    /// Verify a signature over the SignedJsonObject using this public Ed25519
+    /// key.
     ///
-    /// The object must have a signatures key associated  with an object of the
-    /// form `user_id: {key_id: signature}`.
+    /// # Arguments
+    ///
+    /// * `user_id` - The user that claims to have signed this object.
+    ///
+    /// * `key_id` - The ID of the key that was used to sign this object.
+    /// **Note**: The key ID must match the ID of the public key that is
+    /// verifying the signature. This is only used to find the correct
+    /// signature.
+    ///
+    /// * `signed_object` - The signed object that we should check for a valid
+    /// signature.
     ///
     /// Returns Ok if the signature was successfully verified, otherwise an
     /// SignatureError.
+    fn verify_json(
+        &self,
+        user_id: &UserId,
+        key_id: &DeviceKeyId,
+        signed_object: &impl SignedJsonObject,
+    ) -> Result<(), SignatureError>;
+
+    /// Verify a signature over the canonicalized signed JSON object using
+    /// this public Ed25519 key.
     ///
     /// # Arguments
     ///
-    /// * `user_id` - The user who signed the JSON object.
+    /// * `user_id` - The user that claims to have signed this object.
     ///
-    /// * `key_id` - The id of the key that signed the JSON object.
+    /// * `key_id` - The ID of the key that was used to sign this object.
+    /// **Note**: The key ID must match the ID of the public key that is
+    /// verifying the signature. This is only used to find the correct
+    /// signature.
     ///
-    /// * `json` - The JSON object that should be verified.
-    fn verify_json(
+    /// * `canonicalized_json` - The canonicalized version of a signed JSON
+    /// object.
+    ///
+    /// This method should only be used if a signature of an object should be
+    /// checked multiple times and the canonicalization step wants to be done
+    /// only a single time. Prefer the [`VerifyJson::verify_json`] method
+    /// otherwise.
```

Comment:
```suggestion
    /// This method should only be used if an object's signature needs to be
    /// checked multiple times, and you'd like to avoid performing the
    /// canonicalization step each time. Otherwise, prefer the
    /// [`VerifyJson::verify_json`] method
```